src/transaction/log_replication_atomic.cpp

```diff
@@ -367,4 +370,38 @@ namespace cublog
 	  }
       }
   }
+
+  void
+  atomic_replicator::release_locks_by_tranid (cubthread::entry &thread_entry, const TRANID trid)
+  {
+    assert (m_locked_classes.count (trid) > 0);
+
+    auto [begin, end] = m_locked_classes.equal_range (trid);
+    for (auto it = begin; it != end; it++)
+      {
+	lock_unlock_object (&thread_entry, & (it->second), oid_Root_class_oid, SCH_M_LOCK, true);
+      }
+
+    m_locked_classes.erase (trid);
+  }
+
+  void
+  atomic_replicator::acquire_lock (cubthread::entry &thread_entry, const TRANID trid, const OID *classoid)
+  {
+    assert (!OID_ISNULL (classoid) && !OID_ISTEMP (classoid));
+
+    /* TODO:
+     * If a PTS read transaction holds a lock for an extended period without releasing it for the same class,
+     * the replicator could wait too long to acquire the lock.
+     * In such cases, there might be an introduction of a mechanism to abort read transaction that holds lock.
+     */
+
+    if (lock_object (&thread_entry, classoid, oid_Root_class_oid, SCH_M_LOCK, LK_UNCOND_LOCK) != LK_GRANTED)
```

Comment:
Nit:

In debug mode or limited to system parameters, why not add some logic to check the validity of the class_oid passed to LOG_SCHEMA_MODIFICATION_LOCK or the presence of duplicate std::pairs in std::multimap <TRANID, OID> m_locked_classes?

It depends on you.

Reply:
> Nit:
> 
> In debug mode or limited to system parameters, why not add some logic to check the validity of the class_oid passed to LOG_SCHEMA_MODIFICATION_LOCK or the presence of duplicate std::pairs in std::multimap <TRANID, OID> m_locked_classes?
> 
> It depends on you.

I will verify whether duplicate pairs exist. However, I intend to address the validation of classOID when dealing with DDL-related caches. This is because the validation check for OIDs will be included in the process of storing and discarding the cache

Reply:
>  it seems there won't be multiple logs for a single classOID,

Then, how about going with an assert for this?

Reply:
I have set it to check only in debug mode. If duplicate values are found, I have added assert(false)